File: Chronometer/Java/Cli/src/main/java/com/portfolio/chronometer/Stopwatch.java
package com.portfolio.chronometer;

import java.util.ArrayList;
import java.util.List;

public class Stopwatch {
    public enum State {
        STOPPED,
        RUNNING,
        PAUSED
    }

    public interface Clock {
        long currentTimeMillis();
    }

    private final Clock clock;
    private State state = State.STOPPED;
    private long elapsedMillis = 0;
    private long lastStart = 0;
    private final List<Double> laps = new ArrayList<>();

    public Stopwatch() {
        this(System::currentTimeMillis);
    }

    public Stopwatch(Clock clock) {
        this.clock = clock;
    }

    public synchronized State start() {
        if (state == State.RUNNING) {
            return state;
        }
        lastStart = clock.currentTimeMillis();
        state = State.RUNNING;
        return state;
    }

    public synchronized State resume() {
        if (state == State.RUNNING) {
            return state;
        }
        lastStart = clock.currentTimeMillis();
        state = State.RUNNING;
        return state;
    }

    public synchronized double pause() {
        if (state == State.RUNNING) {
            elapsedMillis += clock.currentTimeMillis() - lastStart;
        }
        state = State.PAUSED;
        return round3(elapsedMillis / 1000.0);
    }

    public synchronized double reset() {
        state = State.STOPPED;
        elapsedMillis = 0;
        lastStart = 0;
        laps.clear();
        return 0.0;
    }

    public synchronized double lap() {
        double current = round3(elapsed());
        if (state == State.RUNNING) {
            laps.add(current);
        }
        return current;
    }

    public synchronized List<Double> laps() {
        return new ArrayList<>(laps);
    }

    public synchronized double elapsed() {
        if (state == State.RUNNING) {
            return round3((elapsedMillis + clock.currentTimeMillis() - lastStart) / 1000.0);
        }
        return round3(elapsedMillis / 1000.0);
    }

    public synchronized State state() {
        return state;
    }

    private double round3(double value) {
        return Math.round(value * 1000) / 1000.0;
    }
}

File: Chronometer/Elixir/Web/Phoenix/src/lib/chronometer/stopwatch.ex
defmodule Chronometer.Stopwatch do
  use GenServer

  defstruct running: false, started_at: nil, elapsed: 0.0

  def start_link(_opts \\ []) do
    GenServer.start_link(__MODULE__, %__MODULE__{}, name: __MODULE__)
  end

  def start, do: GenServer.call(__MODULE__, :start)
  def pause, do: GenServer.call(__MODULE__, :pause)
  def resume, do: GenServer.call(__MODULE__, :resume)
  def reset, do: GenServer.call(__MODULE__, :reset)
  def status, do: GenServer.call(__MODULE__, :status)

  @impl true
  def init(state), do: {:ok, state}

  @impl true
  def handle_call(:start, _from, %__MODULE__{running: false} = state) do
    {:reply, {:ok, %{state: "running"}}, %{state | running: true, started_at: monotonic()}}
  end

  def handle_call(:start, _from, %__MODULE__{} = state) do
    {:reply, {:ok, %{state: "running"}}, state}
  end

  @impl true
  def handle_call(:pause, _from, %__MODULE__{running: true} = state) do
    state = %{state | elapsed: current_elapsed(state), running: false, started_at: nil}
    {:reply, {:ok, %{state: "paused", elapsed: round3(state.elapsed)}}, state}
  end

  def handle_call(:pause, _from, state) do
    {:reply, {:ok, %{state: "paused", elapsed: round3(state.elapsed)}}, state}
  end

  @impl true
  def handle_call(:resume, _from, %__MODULE__{running: false} = state) do
    {:reply, {:ok, %{state: "running"}}, %{state | running: true, started_at: monotonic()}}
  end

  def handle_call(:resume, _from, state) do
    {:reply, {:ok, %{state: "running"}}, state}
  end

  @impl true
  def handle_call(:reset, _from, _state) do
    {:reply, {:ok, %{elapsed: 0}}, %__MODULE__{}}
  end

  @impl true
  def handle_call(:status, _from, state) do
    {:reply, {:ok, status_payload(state)}, state}
  end

  defp status_payload(%__MODULE__{running: true} = state) do
    %{state: "running", elapsed: round3(current_elapsed(state))}
  end

  defp status_payload(%__MODULE__{} = state) do
    %{state: "paused", elapsed: round3(state.elapsed)}
  end

  defp current_elapsed(%__MODULE__{running: true, started_at: started_at} = state) do
    state.elapsed + (monotonic() - started_at) / 1000
  end

  defp current_elapsed(%__MODULE__{} = state), do: state.elapsed

  defp monotonic, do: System.monotonic_time(:millisecond)

  defp round3(value), do: Float.round(value, 3)
end
